lib/client.js

```diff
@@ -181,39 +178,56 @@ class Client extends EventEmitter {
             body.pause()
           }
         }
+        const onEnd = () => {
+          this.socket.removeListener('drain', onDrain)
+
+          if (chunked) {
+            this.socket.cork()
+            this.socket.write('\r\n0\r\n', 'ascii')
+          }
+
+          endRequest()
+        }
         const onDrain = () => {
           body.resume()
         }
+        const onFinished = (err) => {
+          freeSocketFinished()
+          freeBodyFinished()
 
-        body.on('data', onData)
-        this.socket.on('drain', onDrain)
-
-        this.socket.uncork()
-        this[kStream].finished(body, (err) => {
-          cleanup()
-          if (err) {
-            // TODO we might want to wait before previous in-flight
-            // requests are finished before destroying
-            if (this.socket) {
-              finishedSocket(this, cb)
-              this.socket.destroy(err)
-            } else {
-              callback(err, null)
-            }
+          if (!err) {
             return
           }
 
-          assert(this.socket)
+          body
+            .removeListener('data', onData)
+            .removeListener('end', onEnd)
 
-          this.socket.removeListener('drain', onDrain)
+          if (typeof body.destroy === 'function') {
+            body.destroy(err)
```

Comment:
I'm a little unsure whether we should forward the error to the request body here or not.

Reply:
@szmarczak you got any thoughts? i.e.

```js
const myBody = ...
await client.request({ ...options, body: myBody })
```

If the request fails would you expect:

- `myBody` to be destroyed and emit the request error.
- `myBody` to be destroyed.
- Do nothing. Caller owns `myBody` life time.

Reply:
Number 1. IMO it should be as close to the pipeline behavior as possible.